pep-0646.rst

```diff
@@ -733,28 +733,166 @@ hopefully enable a thriving ecosystem of tools for analysing and verifying
 shape properties of numerical computing programs.
 
 
-Backwards Compatibility
-=======================
+Grammar Changes
+===============
+
+This PEP would require two grammar changes. Full diffs of ``python.gram``
+and simple tests to confirm correct behaviour are available at
+https://github.com/mrahtz/cpython/commits/pep646-grammar.
+
+Star expressions in indexes
+---------------------------
+
+The first grammar change enables use of star expressions in index operations (that is,
+within square brackets), necessary to support star-unpacking of TypeVarTuples.
+
+Before:
+
+::
+    
+    slices[expr_ty]:
+        | a=slice !',' { a }
+        | a[asdl_expr_seq*]=','.slice+ [','] { _PyAST_Tuple(a, Load, EXTRA) }
+
+After:
+
+::
+    
+    slices[expr_ty]:
+        | a=slice !',' { a }
+        | a[asdl_expr_seq*]=','.(slice | starred_expression)+ [','] { _PyAST_Tuple(a, Load, EXTRA) }
+
+As with star-unpacking in other contexts, the star operator would call ``__iter__`` on the callee,
+and add the contents of the resulting iterator to the list of index arguments.
+This would enable ``TypeVarTuple`` to be implemented as roughly:
+
+::
 
-In order to use the star operator for unpacking of ``TypeVarTuple`` instances,
-we would need to make two grammar changes:
+    class TypeVarTuple:
+        def __init__(self, name):
+            self._unpacked = _UnpackedTypeVarTuple(name) 
+        def __iter__(self):
+            yield self._unpacked
 
-1. Star expressions must be made valid in at least index operations.
-   For example, ``Tuple[*Ts]`` and ``Tuple[T1, *Ts, T2]`` would both
-   be valid. (This PEP does not allow multiple unpacked ``TypeVarTuple``
-   instances to appear in a single parameter list, so ``Tuple[*Ts1, *Ts2]``
-   would be a runtime error. Also note that star expressions would *not*
-   be valid in slice expressions -  e.g. ``Tuple[*Ts:*Ts]`` is
-   nonsensical and should remain invalid.)
-2. We would need to make '``*args: *Ts``' valid in function definitions.
+There are three things to note about this grammar change. First, for consistency with
+star-unpacking in other contexts, ``_UnpackedTypeVarTuple`` would always be received
+by ``__class_getitem__`` as part of a tuple:
```

Comment:
To describe the semantics of this grammar, we need neither `_UnpackedTypeVarTuple` nor `__class_getitem__`. Rather, we need to describe what the instance method `__getitem__` will receive when the argument has various forms involving `*`.

Fortunately it seems to be that in all cases where the substitution is valid, `x[..., *a, ...]` produces the same as `x[(..., *a, ...)]`, with the one edge case that `x[*a]` becomes `x[(*a,)]`. Of course if the dots involve a slice, the substitution is not valid, but we can then replace slices with calls to `slice(...)` to make things valid, for example `x[1:2, *a]` will become `x[(slice(1, 2), *a)]`. And then we can just defer to the existing semantics for `*a` and for `x[y]`.

We probably will need to defend why we even allow `x[1:2, *a]`, since we don't need it for variadic types. Our case is somewhat weak here (and we can state that we'd be fine if the SC were to strike this), but to me it seems that it's the simplest grammar change. After all that's also why we chose to allow e.g. `x[*a, *b]`.

Reply:
Mm, great points. I've rewritten this section to explain these points (incorporating your wording, which I thought was very clear). What do you think?